source/base/mpi.cc

```diff
@@ -741,8 +805,14 @@ namespace Utilities
       // would run after MPI_Finalize is called, leading to errors
 
 #ifdef DEAL_II_WITH_MPI
-      // Start with the deal.II MPI vectors (need to do this before finalizing
-      // PETSc because it finalizes MPI).  Delete vectors from the pools:
+      // Before exiting, wait for nonblocking communication to complete:
+      for (auto request : requests)
+        {
+          const int ierr = MPI_Wait(request, MPI_STATUS_IGNORE);
```

Comment:
Wouldn't you just want to call `MPI_Request_free()` on the requests? I can imagine situations where communication wasn't finished for some reason, and this leading to a deadlock.

Reply:
I am not sure about that. A wait might hang, but what happens in a regular program when I free and immediately MPI_finalize if the communication is still running?

Reply:
I'm thinking of the case where one processor has called `MPI_Abort` and won't send anything. I think that canceling a `Request` object shouldn't be controversial: You're not canceling the communication, just the object that exposes the communication to the user side.